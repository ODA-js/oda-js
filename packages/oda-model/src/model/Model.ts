import { Map, Record } from 'immutable';

import { IModel, IModelInit, IModelStore, IModelTransform } from '../interfaces/IModel';
import { IPackage, IPackageInit } from '../interfaces/IPackage';
import { Package } from './Package';
import { Persistent } from './Persistent';

// tslint:disable-next-line:variable-name
export const DefaultModel: IModelStore = {
  name: null,
  title: null,
  description: null,
  packages: null,
};

// tslint:disable-next-line:variable-name
export const ModelTransform: IModelTransform = {
  packages: {
    transform: (input: IPackageInit[], model: IModel) => {
      if (input) {
        return Map<string, IPackage>(input.map(p => [p.name, new Package({
          ...p,
          model,
        },
        )]) as [string, IPackage][]);
      } else {
        return null;
      }
    },
    reverse: (input?: Map<string, IPackage>) => {
      if (input) {
        return Array.from(input.values()[Symbol.iterator]()).map(f => f.toJS() as IPackageInit);
      } else {
        return null;
      }
    },
  },
};

// tslint:disable-next-line:variable-name
export const ModelStorage = Record(DefaultModel);

export class Model extends Persistent<IModelInit, IModelStore> implements IModel {
  public get modelType(): 'model' {
    return 'model';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get packages(): Map<string, IPackage> {
    return this.store.get('packages', null);
  }
  public get defaultPackage(): IPackage {
    return this.packages.get(this.defaultPackageName);
  }

  public get defaultPackageName(): string {
    return this._defaultPackageName;
  }

  private _defaultPackageName: string = 'system';

  protected transform(input: Partial<IModelInit>): IModelStore {
    const result: IModelStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'packages') {
            result.packages = ModelTransform.packages.transform(input.packages, this);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IModelStore> & Readonly<IModelStore>): IModelInit {
    const result: IModelInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'packages') {
            result.packages = ModelTransform.packages.reverse(input.packages);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init: Partial<IModelInit> = {
    defaultPackageName: 'system',
  }) {
    super();

    if (init.hasOwnProperty('defaultPackageName') && init.defaultPackageName && init.defaultPackageName !== this.defaultPackageName) {
      this._defaultPackageName = init.defaultPackageName;
    }
    this.store = new ModelStorage(this.transform(init));
    this.init = new (Record<Partial<IModelInit>>(init))();
    if (!this.packages || !this.packages.has(this.defaultPackageName)) {
      this.updateWith({
        packages: [{
          acl: 0,
          name: this.defaultPackageName,
          items: [],
          model: this,
        }],
      });
    }
  }
}
