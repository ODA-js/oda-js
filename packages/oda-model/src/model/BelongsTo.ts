import { Record } from 'immutable';

import { IBelongsTo, IBelongsToInit, IBelongsToStore, IRelationTransform } from '../interfaces/IBelongsTo';
import { IEntityRef } from '../interfaces/IEntityRef';
import { Relation } from './Relation';
import { TransformField, TransformRef } from './utils';

// tslint:disable-next-line:variable-name
export const DefaultBelongsTo: IBelongsToStore = {
  name: null,
  title: null,
  description: null,
  fields: null,
  opposite: null,
  belongsTo: null,
  //storage
  single: true,
  stored: true,
  embedded: true,
  // name
  fullName: null,
  normalName: null,
  shortName: null,
};

// tslint:disable-next-line:variable-name
export const BelongsToTransform: IRelationTransform = {
  belongsTo: TransformRef(),
  fields: TransformField(),
};

// tslint:disable-next-line:variable-name
export const BelongsToStorage = Record(DefaultBelongsTo);

export class BelongsTo extends Relation<IBelongsToInit, IBelongsToStore> implements IBelongsTo {
  public get verb(): 'BelongsTo' {
    return 'BelongsTo';
  }
  public get ref(): IEntityRef {
    return this.store.get('belongsTo', null);
  }
  public get belongsTo(): IEntityRef {
    return this.store.get('belongsTo', null);
  }

  protected transform(input: Partial<IBelongsToInit>): Partial<IBelongsToStore> {
    const result: Partial<IBelongsToStore> = {};
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'belongsTo') {
            result.belongsTo = BelongsToTransform.belongsTo.transform(input.belongsTo);
          } else if (f === 'fields') {
            result.fields = BelongsToTransform.fields.transform(input.fields);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IBelongsToStore> & Readonly<IBelongsToStore>): Partial<IBelongsToInit> {
    const result: Partial<IBelongsToInit> = {};
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'belongsTo' || f === 'ref') {
            result.belongsTo = BelongsToTransform.belongsTo.reverse(input.belongsTo);
          } else if (f === 'fields') {
            result.fields = BelongsToTransform.fields.reverse(input.fields);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init: Partial<IBelongsToInit> = {}) {
    super();
    this.store = new BelongsToStorage(this.transform(init));
    this.init = new (Record<Partial<IBelongsToInit>>(init))();
  }
}
