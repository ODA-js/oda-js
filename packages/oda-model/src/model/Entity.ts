import { Map, Record, Set } from 'immutable';

import { IEntity, IEntityInit, IEntityStore, IEntityTransform, IEntityACL} from '../interfaces/IEntity';
import { IField } from '../interfaces/IField';
import { IPackage } from '../interfaces/IPackage';
import { Persistent } from './Persistent';
import { TransformField } from './utils';
import { IPackageContext } from '../contexts/IPackageContext';
import { IEntityContext } from '../contexts/IEntityContext';

// tslint:disable-next-line:variable-name
export const DefaultEntity: IEntityStore = {
  name: null,
  title: null,
  description: null,
  acl: null,
  plural: null,
  singular: null,
  fields: null,
  storage: null,
  indexed: null,
  relations: null,
  required: null,
};

// tslint:disable-next-line:variable-name
export const EntityTransform: IEntityTransform = {
  fields: TransformField(),
};

// tslint:disable-next-line:variable-name
export const EntityStorage = Record(DefaultEntity);

export class Entity extends Persistent<IEntityInit, IEntityStore, IPackageContext> implements IEntity {
  public get modelType(): 'entity' {
    return 'entity';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get acl() {
    return this.store.get('acl', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get singular(): string {
    return this.store.get('singular', null);
  }
  public get plural(): string {
    return this.store.get('plural', null);
  }
  public get fields(): Map<string, IField> {
    return this.store.get('fields', null);
  }
  public get relations(): Set<string> {
    return this.store.get('relations', null);
  }
  public get required(): Set<string> {
    return this.store.get('required', null);
  }
  public get indexed(): Set<string> {
    return this.store.get('indexed', null);
  }
  protected transform(input: Partial<IEntityInit>): IEntityStore {
    const result: IEntityStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'fields') {
            result.fields = EntityTransform.fields.transform(input.fields, this);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IEntityStore> & Readonly<IEntityStore>): IEntityInit {
    const result: IEntityInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'fields') {
            result.fields = EntityTransform.fields.reverse(input.fields);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init?: Partial<IEntityInit>, context?: IPackageContext) {
    super();
    if (!context && init && init.fields) {
      throw new Error('context must be provided');
    }
    if (context) {
      this.attach(context);
    }
    this.store = new EntityStorage(this.transform(init));
  }
}
