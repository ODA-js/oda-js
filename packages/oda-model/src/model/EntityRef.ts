import camelcase from 'camelcase';
import { Record } from 'immutable';
import * as inflected from 'inflected';

import { DEFAULT_ID_FIELDNAME, REF_PATTERN } from '../definitions';
import { IEntityRef, IEntityRefInit, IEntityRefStore } from '../interfaces/IEntityRef';
import { Persistent } from './Persistent';
import { IRelationContext } from '../contexts/IRelationContext';

// tslint:disable-next-line:variable-name
export const DefaultEntityRef: IEntityRefStore = {
  backField: '',
  entity: '',
  field: '',
};

// tslint:disable-next-line:variable-name
export const EntityRefStorage = Record(DefaultEntityRef);

export class EntityRef extends Persistent<IEntityRefInit, IEntityRefStore, IRelationContext>
  implements IEntityRef {
  public get backField(): string {
    return this.store.get('backField', '');
  }
  public get entity(): string {
    return this.store.get('entity', '');
  }
  public get field(): string {
    return this.store.get('field', '');
  }

  public toString(): string {
    return `${this.backField ? (this.backField + '@') : ''}${this.entity}#${this.field || DEFAULT_ID_FIELDNAME}`;
  }

  protected transform(input: Partial<IEntityRef>): IEntityRef {
    const result: IEntityRef = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          result[f] = input[f];
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IEntityRefStore> & Readonly<IEntityRefStore>): IEntityRef {
    const result: IEntityRef = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          result[f] = core[f];
        }
      }
    }
    return result;
  }

  constructor(init: string | Partial<IEntityRef> = {}) {
    super();
    if (typeof init === 'string') {
      let res = init.match(REF_PATTERN);
      if (res && res.length > 0) {
        this.store = new EntityRefStorage(this.transform({
          backField: res[1],
          entity: inflected.classify(res[2]),
          field: camelcase(res[3].trim()),
        }));
      }
    } else {
      this.store = new EntityRefStorage(this.transform(init));
    }
  }
}
