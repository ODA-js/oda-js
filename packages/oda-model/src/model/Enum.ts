import { Record } from 'immutable';
import { Map } from 'immutable';

import { IEnumStore, IEnumTransform, IEnumItem, IEnumInit, IEnum, EnumInitItem } from '../interfaces/IEnum';
import { transformMap } from './utils';
import { Persistent } from './Persistent';
import { } from './interfaces/IEnum';
import { IFieldStore } from '../interfaces/IField';

// tslint:disable-next-line:variable-name
export const DefaultEnum: Partial<IEnumStore> = {
  name: null,
  title: null,
  description: null,
  package: null,
  values: null,
};

// tslint:disable-next-line:variable-name
export const EnumTransform: IEnumTransform = {
  values: {
    transform: (input: EnumInitItem[] | {
      [name: string]: EnumInitItem;
    }) => {
      if (!Array.isArray(input)) {
        input = Object.keys(input).map(k => input[k]);
      }
      return Map<string, IEnumItem>(input.map(p => {
        if (typeof p === 'string') {
          return [p, {
            value: p,
          }];
        } else {
          return [p, p];
        }
      }) as [string, IEnumItem][]);
    },
    reverse: (input: Map<string, IEnumItem>) => Array.from(input.values()[Symbol.iterator]()),
  },
};

// tslint:disable-next-line:variable-name
export const EnumStorage = Record(DefaultEnum);

export class Enum extends Persistent<IEnumInit, IEnumStore> implements IEnum {
  public get modelType(): 'enum' {
    return 'enum';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get values(): Map<string, IEnumItem> {
    return this.store.get('values', null);
  }
  protected transform(input: Partial<IEnumInit>): IEnumStore {
    const result: IEnumStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'values') {
            result.values = EnumTransform.values.transform(input.values);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }
  protected reverse(input: IEnumStore): IEnumInit {
    const result: IEnumInit = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'values') {
            result.values = EnumTransform.values.reverse(input.values);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init: Partial<IEnumInit> = {}) {
    super();
    this.store = new EnumStorage(this.transform(init));
    this.init = new (Record<Partial<IEnumInit>>(init))();
  }
}
