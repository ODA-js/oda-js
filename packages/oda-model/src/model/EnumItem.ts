import { Map, Record } from 'immutable';

import { IEnum, IEnumInit, IEnumStore, IEnumTransform } from '../interfaces/IEnum';
import { Persistent } from './Persistent';
import { IEnumItem, IEnumItemInit } from '../interfaces/IEnumItem';

// tslint:disable-next-line:variable-name
export const DefaultEnumItem: IEnumItemInit = {
  name: null,
  title: null,
  description: null,
  type: null,
  value: null,
};

// tslint:disable-next-line:variable-name
export const EnumStorage = Record(DefaultEnumItem);

export class EnumItem extends Persistent<IEnumItemInit, IEnumItemInit> implements IEnumItem {
  public get modelType(): 'enum' {
    return 'enum';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get value(): string {
    return this.store.get('value', null);
  }
  protected transform(input: Partial<IEnumInit>): IEnumStore {
    const result: IEnumStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          result[f] = input[f];
        }
      }
    }
    return result;
  }
  protected reverse(input: Record<IEnumStore> & Readonly<IEnumStore>): IEnumInit {
    const result: IEnumInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          result[f] = core[f];
        }
      }
    }
    return result;
  }

  constructor(init: Partial<IEnumItemInit> = {}) {
    super();
    this.store = new EnumStorage(this.transform(init));
    this.init = new (Record<Partial<IEnumItemInit>>(init))();
  }
}
