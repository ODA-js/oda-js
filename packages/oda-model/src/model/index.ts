import { Entity } from './entity';
import { EntityBase } from './entitybase';
import { Mixin } from './mixin';
import { Union } from './union';
import { EntityReference } from './entityreference';
import { Field } from './field';
import { FieldBase } from './fieldbase';
import { HasOne } from './hasone';
import { HasMany } from './hasmany';
import { BelongsTo } from './belongsto';
import { BelongsToMany } from './belongstomany';
import { MetaModel } from './metamodel';
import { ModelBase } from './modelbase';
import { Mutation } from './mutation';
import { DEFAULT_ID_FIELD } from './definitions';
import { ModelPackage } from './modelpackage';
import { RelationBase } from './relationbase';
import {
  FieldArgs, MetaModelStore,
  ModelHook, IValidationResult, ValidationResultType,
} from './interfaces';
import { Metadata } from './metadata';
import Validator from './validator/index';

export {
  ModelHook,
  Metadata,
  FieldArgs,
  MetaModelStore,
  Entity,
  Field,
  HasOne,
  HasMany,
  BelongsTo,
  BelongsToMany,
  ModelPackage,
  MetaModel,
  DEFAULT_ID_FIELD,
  Mutation,
  RelationBase,
  FieldBase,
  EntityReference,
  ModelBase,
  IValidationResult,
  ValidationResultType,
  Validator,
};
