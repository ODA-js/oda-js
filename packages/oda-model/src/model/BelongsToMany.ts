import { Map, Record } from 'immutable';

import { IBelongsToMany, IBelongsToManyInit, IBelongsToManyStore } from '../interfaces/IBelongsToMany';
import { IEntityRef } from '../interfaces/IEntityRef';
import { IField, IFieldInit } from '../interfaces/IField';
import { EntityRef } from './EntityRef';
import { Field } from './Field';
import { Relation } from './Relation';


// tslint:disable-next-line:variable-name
export const DefaultBelongsToMany: IBelongsToManyStore = {
  name: null,
  title: null,
  description: null,
  fields: null,
  opposite: null,
  belongsToMany: null,
  using: null,
  //storage
  single: false,
  stored: false,
  embedded: false,
  // name
  fullName: null,
  normalName: null,
  shortName: null,
};

// tslint:disable-next-line:variable-name
export const BelongsToManyTransform = {
  belongsToMany: {
    transform: (inp) => {
      if (inp) {
        return new EntityRef(inp);
      } else {
        return null;
      }
    },
    reverse: (inp) => {
      if (inp) {
        return inp.toString();
      } else {
        return null;
      }
    },
  },
  fields: {
    transform: (input: IFieldInit[]) => {
      if (input) {
        return Map<string, IField>(input.map(p => [p.name, new Field(p)]) as [string, IField][]);
      } else {
        return null;
      }
    },
    reverse: (input: Map<string, IField>) => {
      if (input) {
        return Array.from(input.values()[Symbol.iterator]()).map(i => i.toJS());
      } else {
        return null;
      }
    },
  },
  using: {
    transform: (inp) => {
      if (inp) {
        return new EntityRef(inp);
      } else {
        return null;
      }
    },
    reverse: (inp) => {
      if (inp) {
        return inp.toString();
      } else {
        return null;
      }
    },
  },
};

// tslint:disable-next-line:variable-name
export const BelongsToManyStorage = Record(DefaultBelongsToMany);

export class BelongsToMany
  extends Relation<IBelongsToManyInit, IBelongsToManyStore> implements IBelongsToMany {
  public get verb(): 'BelongsToMany' {
    return 'BelongsToMany';
  }
  public get ref(): IEntityRef {
    return this.store.get('belongsToMany', null);
  }
  public get belongsToMany(): IEntityRef {
    return this.store.get('belongsToMany', null);
  }
  public get using(): IEntityRef {
    return this.store.get('using', null);
  }

  protected transform(input: Partial<IBelongsToManyInit>): Partial<IBelongsToManyStore> {
    const result: IBelongsToManyStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'belongsToMany') {
            result.belongsToMany = BelongsToManyTransform.belongsToMany.transform(input.belongsToMany);
          } else if (f === 'using') {
            result.using = BelongsToManyTransform.using.transform(input.using);
          } else if (f === 'fields') {
            result.fields = BelongsToManyTransform.fields.transform(input.fields);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IBelongsToManyStore> & Readonly<IBelongsToManyStore>): Partial<IBelongsToManyInit> {
    const result: IBelongsToManyInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'belongsToMany' || f === 'ref') {
            result.belongsToMany = BelongsToManyTransform.belongsToMany.reverse(input.belongsToMany);
          } else if (f === 'using') {
            result.using = BelongsToManyTransform.using.reverse(input.using);
          } else if (f === 'fields') {
            result.fields = BelongsToManyTransform.fields.reverse(input.fields);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init: Partial<IBelongsToManyInit> = {}) {
    super();
    this.store = new BelongsToManyStorage(this.transform(init));
    this.init = new (Record<Partial<IBelongsToManyInit>>(init))();
  }
}
