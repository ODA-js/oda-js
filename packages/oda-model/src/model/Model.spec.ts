import { Model } from './Model';
import { ModelFactory } from './../model/Factory';
import schema from '../01_test_schema/';

describe('Model', () => {
  let model: Model;

  beforeAll(() => {
    expect(() => new Model()).not.toThrow();
  });

  beforeEach(() => {
    model = new Model();
  });

  it('Model.load', () => {
    expect(() => ModelFactory.createModel(schema)).not.toThrow();
  });

  it('load serialized model', () => {
    let serialized;
    expect(() => serialized = ModelFactory.createModel(schema)).not.toThrow();
    expect(serialized).not.toBeUndefined();
    expect(serialized).toMatchSnapshot();
    expect(serialized.packages.size).toBe(3);
    expect(serialized.packages.has('system')).toBeTruthy();
    expect(serialized.toJS()).toMatchSnapshot();
  });

  it('create empty', () => {
    expect(model.name).toBeNull();
    expect(model.modelType).toBe('model');
    expect(model.description).toBeNull();
    expect(model.packages.size).toBe(1);
    expect(model.defaultPackage).not.toBeUndefined();
    expect(model.defaultPackage.toJS()).toMatchObject({
      name: 'system',
    });
  });

  it('update strings', () => {
    expect(() => model.updateWith({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
    })).not.toThrow();

    expect(model.toJS()).toMatchObject({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
    });
  });

  it('update packages', () => {
    expect(() => model.updateWith({
      packages: [],
    })).not.toThrow();
    expect(model.defaultPackage).not.toBeNull();
  });

  it('toJS with dupes', () => {
    expect(() => model.updateWith({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
      packages: [
        { name: 'one', acl: 100, items: [] },
        { name: 'one', acl: 100 , items: []},
      ],
    })).not.toThrow();

    expect(model.packages.size).toBe(2);

    expect(model.toJS()).toMatchObject({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
      packages: [
        { name: 'system' },
        { name: 'one' },
      ],
    });
  });
});
