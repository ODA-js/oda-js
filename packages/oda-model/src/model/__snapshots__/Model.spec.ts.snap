// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model create empty 1`] = `
Model {
  "_defaultPackageName": "system",
  "store": Immutable.Record {
    "name": null,
    "title": null,
    "description": null,
    "packages": Immutable.Map {
      "system": Package {
        "_context": Immutable.Record {
          "model": [Circular],
        },
        "store": Immutable.Record {
          "name": "system",
          "title": null,
          "description": null,
          "acl": 0,
          "abstract": null,
          "items": Immutable.Map {},
        },
      },
    },
  },
}
`;

exports[`Model create empty 2`] = `
Object {
  "packages": Array [
    Object {
      "acl": 0,
      "items": Array [],
      "name": "system",
    },
  ],
}
`;

exports[`Model load serialized model 1`] = `
Object {
  "name": "ToDoListApplication",
  "packages": Array [
    Object {
      "acl": 0,
      "items": Array [
        Object {
          "description": "User",
          "fields": Array [
            Object {
              "identity": true,
              "name": "userName",
            },
            Object {
              "name": "password",
              "required": true,
            },
            Object {
              "name": "isSystem",
              "type": "boolean",
            },
            Object {
              "name": "enabled",
              "type": "boolean",
            },
            Object {
              "name": "groups",
              "relation": Object {
                "belongsToMany": "Group#id",
                "fields": null,
                "using": "UsersToGroupsMap#user",
              },
            },
          ],
          "name": "User",
        },
        Object {
          "description": "Group of users that can share common ToDos",
          "fields": Array [
            Object {
              "identity": true,
              "indexed": "text",
              "name": "name",
            },
            Object {
              "indexed": "text",
              "name": "description",
            },
            Object {
              "name": "users",
              "relation": Object {
                "belongsToMany": "User#id",
                "fields": null,
                "using": "UsersToGroupsMap#group",
              },
            },
            Object {
              "name": "todoItems",
              "relation": Object {
                "belongsToMany": "ToDoItem#id",
                "fields": null,
                "using": "ToDoItemsSharedToGroupsMap#group",
              },
            },
          ],
          "name": "Group",
        },
        Object {
          "description": "Users to Groups mapping",
          "fields": Array [
            Object {
              "identity": "linkTable",
              "name": "user",
            },
            Object {
              "identity": "linkTable",
              "name": "group",
            },
            Object {
              "name": "userLink",
              "relation": Object {
                "belongsTo": "user@User#id",
                "fields": null,
              },
            },
            Object {
              "name": "groupLink",
              "relation": Object {
                "belongsTo": "group@Group#id",
                "fields": null,
              },
            },
          ],
          "name": "UsersToGroupsMap",
        },
        Object {
          "description": "ToDo item",
          "fields": Array [
            Object {
              "name": "sharedTo",
              "relation": Object {
                "belongsToMany": "Group#id",
                "fields": null,
                "using": "ToDoItemsSharedToGroupsMap#item",
              },
            },
            Object {
              "name": "done",
              "type": "Boolean",
            },
            Object {
              "indexed": true,
              "name": "assignedTo",
              "relation": Object {
                "belongsTo": "User#id",
                "fields": null,
              },
            },
            Object {
              "indexed": "text",
              "name": "name",
            },
            Object {
              "indexed": true,
              "name": "creator",
              "relation": Object {
                "belongsTo": "User#id",
                "fields": null,
              },
            },
            Object {
              "name": "dueTo",
              "type": "Date",
            },
            Object {
              "name": "isPrivate",
              "type": "Boolean",
            },
            Object {
              "name": "type",
              "type": "ToDotype",
            },
            Object {
              "indexed": "text",
              "name": "description",
            },
          ],
          "name": "ToDoItem",
        },
        Object {
          "description": "Sharing ToDo items between groups",
          "fields": Array [
            Object {
              "identity": "shared",
              "indexed": true,
              "name": "item",
            },
            Object {
              "identity": "shared",
              "indexed": true,
              "name": "group",
            },
            Object {
              "name": "itemLink",
              "relation": Object {
                "belongsTo": "item@ToDoItem#id",
                "fields": null,
              },
            },
            Object {
              "name": "groupLink",
              "relation": Object {
                "belongsTo": "group@Group#id",
                "fields": null,
              },
            },
          ],
          "name": "ToDoItemsSharedToGroupsMap",
        },
        Object {
          "acl": Object {
            "execute": Array [],
          },
          "args": Array [
            Object {
              "name": "userName",
              "required": true,
            },
            Object {
              "name": "password",
              "required": true,
            },
          ],
          "description": "make user login",
          "name": "loginUser",
          "payload": Array [
            Object {
              "name": "token",
              "type": "string",
            },
          ],
        },
        Object {
          "name": "ToDoType",
          "values": Array [
            Object {
              "name": "usual",
            },
            Object {
              "name": "super pupper",
              "type": "string",
              "value": "super",
            },
          ],
        },
        Object {
          "name": "UserKind",
          "values": Array [
            Object {
              "name": "public",
            },
            Object {
              "name": "registered",
            },
            Object {
              "name": "super pupper",
              "title": "The super power User",
              "type": "string",
              "value": "admin",
            },
          ],
        },
      ],
      "name": "system",
    },
    Object {
      "abstract": true,
      "acl": 100,
      "items": Array [
        Object {
          "description": "User",
          "fields": Array [
            Object {
              "identity": true,
              "name": "userName",
            },
            Object {
              "name": "password",
              "required": true,
            },
            Object {
              "name": "isSystem",
              "type": "boolean",
            },
            Object {
              "name": "enabled",
              "type": "boolean",
            },
            Object {
              "name": "groups",
              "relation": Object {
                "belongsToMany": "Group#id",
                "fields": null,
                "using": "UsersToGroupsMap#user",
              },
            },
          ],
          "name": "User",
        },
        Object {
          "description": "Group of users that can share common ToDos",
          "fields": Array [
            Object {
              "identity": true,
              "indexed": "text",
              "name": "name",
            },
            Object {
              "indexed": "text",
              "name": "description",
            },
            Object {
              "name": "users",
              "relation": Object {
                "belongsToMany": "User#id",
                "fields": null,
                "using": "UsersToGroupsMap#group",
              },
            },
            Object {
              "name": "todoItems",
              "relation": Object {
                "belongsToMany": "ToDoItem#id",
                "fields": null,
                "using": "ToDoItemsSharedToGroupsMap#group",
              },
            },
          ],
          "name": "Group",
        },
        Object {
          "description": "ToDo item",
          "fields": Array [
            Object {
              "name": "sharedTo",
              "relation": Object {
                "belongsToMany": "Group#id",
                "fields": null,
                "using": "ToDoItemsSharedToGroupsMap#item",
              },
            },
            Object {
              "name": "done",
              "type": "Boolean",
            },
            Object {
              "indexed": true,
              "name": "assignedTo",
              "relation": Object {
                "belongsTo": "User#id",
                "fields": null,
              },
            },
            Object {
              "indexed": "text",
              "name": "name",
            },
            Object {
              "indexed": true,
              "name": "creator",
              "relation": Object {
                "belongsTo": "User#id",
                "fields": null,
              },
            },
            Object {
              "name": "dueTo",
              "type": "Date",
            },
            Object {
              "name": "isPrivate",
              "type": "Boolean",
            },
            Object {
              "name": "type",
              "type": "ToDotype",
            },
            Object {
              "indexed": "text",
              "name": "description",
            },
          ],
          "name": "ToDoItem",
        },
        Object {
          "description": "Sharing ToDo items between groups",
          "fields": Array [
            Object {
              "identity": "shared",
              "indexed": true,
              "name": "item",
            },
            Object {
              "identity": "shared",
              "indexed": true,
              "name": "group",
            },
            Object {
              "name": "itemLink",
              "relation": Object {
                "belongsTo": "item@ToDoItem#id",
                "fields": null,
              },
            },
            Object {
              "name": "groupLink",
              "relation": Object {
                "belongsTo": "group@Group#id",
                "fields": null,
              },
            },
          ],
          "name": "ToDoItemsSharedToGroupsMap",
        },
        Object {
          "description": "Users to Groups mapping",
          "fields": Array [
            Object {
              "identity": "linkTable",
              "name": "user",
            },
            Object {
              "identity": "linkTable",
              "name": "group",
            },
            Object {
              "name": "userLink",
              "relation": Object {
                "belongsTo": "user@User#id",
                "fields": null,
              },
            },
            Object {
              "name": "groupLink",
              "relation": Object {
                "belongsTo": "group@Group#id",
                "fields": null,
              },
            },
          ],
          "name": "UsersToGroupsMap",
        },
      ],
      "name": "all",
    },
    Object {
      "abstract": false,
      "acl": 0,
      "items": Array [
        Object {
          "description": "ToDo item",
          "fields": Array [
            Object {
              "name": "sharedTo",
              "relation": Object {
                "belongsToMany": "Group#id",
                "fields": null,
                "using": "ToDoItemsSharedToGroupsMap#item",
              },
            },
            Object {
              "name": "done",
              "type": "Boolean",
            },
            Object {
              "indexed": true,
              "name": "assignedTo",
              "relation": Object {
                "belongsTo": "User#id",
                "fields": null,
              },
            },
            Object {
              "indexed": "text",
              "name": "name",
            },
            Object {
              "indexed": true,
              "name": "creator",
              "relation": Object {
                "belongsTo": "User#id",
                "fields": null,
              },
            },
            Object {
              "name": "dueTo",
              "type": "Date",
            },
            Object {
              "name": "isPrivate",
              "type": "Boolean",
            },
            Object {
              "name": "type",
              "type": "ToDotype",
            },
            Object {
              "indexed": "text",
              "name": "description",
            },
          ],
          "name": "ToDoItem",
        },
        Object {
          "acl": Object {
            "execute": Array [],
          },
          "args": Array [
            Object {
              "name": "userName",
              "required": true,
            },
            Object {
              "name": "password",
              "required": true,
            },
          ],
          "description": "make user login",
          "name": "loginUser",
          "payload": Array [
            Object {
              "name": "token",
              "type": "string",
            },
          ],
        },
      ],
      "name": "public",
    },
  ],
}
`;

exports[`Model toJS with dupes 1`] = `
Object {
  "description": "very cool",
  "name": "cool",
  "packages": Array [
    Object {
      "acl": 0,
      "items": Array [],
      "name": "system",
    },
    Object {
      "acl": 100,
      "items": Array [],
      "name": "one",
    },
  ],
  "title": "very cool title",
}
`;

exports[`Model update packages 1`] = `
Object {
  "packages": Array [
    Object {
      "acl": 0,
      "items": Array [],
      "name": "system",
    },
  ],
}
`;

exports[`Model update strings 1`] = `
Object {
  "description": "very cool",
  "name": "cool",
  "packages": Array [
    Object {
      "acl": 0,
      "items": Array [],
      "name": "system",
    },
  ],
  "title": "very cool title",
}
`;

exports[`Model updates with null or undefined 1`] = `
Object {
  "description": "very cool",
  "name": "cool",
  "packages": Array [
    Object {
      "acl": 0,
      "items": Array [],
      "name": "system",
    },
  ],
  "title": "very cool title",
}
`;

exports[`Model updates with null or undefined 2`] = `
Object {
  "name": "cool",
  "packages": Array [
    Object {
      "acl": 0,
      "items": Array [],
      "name": "system",
    },
  ],
}
`;
